lib/senders/email.js

```diff
@@ -253,97 +267,227 @@ module.exports = function (log, config) {
+  // satisfy both criteria to count as a match. Where an email address matches
+  // both sendgrid and ses, sendgrid wins.
+  //
+  // @param {Object} message
+  //
+  // @returns {Promise} Resolves to an array of `Service` objects.
+  //
+  // @typedef {Object} Service
+  //
+  // @property {Object} mailer           The object on which to invoke the `sendMail`
+  //                                     method.
+  //
+  // @property {String[]} emailAddresses The array of email addresses to send to.
+  //                                     The address at index 0 will be used as the
+  //                                     `to` address and any remaining addresses
+  //                                     will be included as `cc` addresses.
+  //
+  // @property {String} emailService     The name of the email service for metrics.
+  //
+  // @property {String} emailSender      The name of the underlying email sender,
+  //                                     used for both metrics and sent as the
+  //                                     `provider` param in external requests.
+  Mailer.prototype.selectEmailServices = function (message) {
+    const mailers = {
+      internal: this.mailer,
+      external: this.emailService
+    }
+    const indices = new Map()
+    const emailAddresses = [ message.email ]
+    if (Array.isArray(message.ccEmails)) {
+      emailAddresses.push(...message.ccEmails)
+    }
+
+    return redis.get('config')
+      .catch(() => {})
+      .then(liveConfig => {
+        return emailAddresses.reduce((services, emailAddress) => {
+          if (liveConfig) {
+            const { sendgrid, ses } = liveConfig
+
+            if (sendgrid && isLiveConfigMatch(sendgrid, emailAddress)) {
+              return upsertServicesArrayExternalSendgrid(services, emailAddress)
             }
-          )
-          return d.reject(err)
-        }
-        log.info(
-          {
-            op: 'mailer.send.1',
-            status: status && status.message,
-            id: status && status.messageId,
-            to: emailConfig && emailConfig.to
+
+            if (ses && isLiveConfigMatch(ses, emailAddress)) {
+              return upsertServicesArrayExternalSes(services, emailAddress)
+            }
+          }
+
+          if (config.emailService.forcedEmailAddresses.test(message.email)) {
+            return upsertServicesArrayExternalSes(services, emailAddress)
           }
-        )
 
-        emailUtils.logEmailEventSent(log, message)
+          return upsertServicesArrayInternal(services, emailAddress)
+        }, [])
+      })
+
+    function isLiveConfigMatch (liveConfig, emailAddress) {
+      const { percentage, regex } = liveConfig
+
+      if (percentage >= 0 && percentage < 100) {
+        let sum = 0
 
-        return d.resolve(status)
+        for (const character in emailAddress) {
+          sum += emailAddress.charCodeAt(character)
+        }
+
+        if (sum % 100 >= percentage) {
```

Comment:
I wonder if there would be any clustering around particular values due to things like domain name?

Reply:
I see if I can define a representative data sample and then run some tests on that.

Reply:
I tried to run a test for this using only `mozilla.com` email addresses. My thinking was that enforcing a single domain name in the test data it would be the best check for clustering. The resulting distribution was not as flat as I was hoping, but then I compared it to the distribution returned by our existing `features.isSampledUser` function and that was also not as flat as I was hoping. You can see those distributions in [this gist](https://gist.github.com/philbooth/8bc1d72a4885b17ef6e6f4061c4ccb95).

So really I need to do it with a much bigger data set. I'll have to write a script to scrape email addresses from phonebook or something and I haven't got time to do that right now. I already spent too long on this today, there's other things I was supposed to be doing and am behind with now. Hence moving this back to `active` and sticking a `WIP` on it, sadly.